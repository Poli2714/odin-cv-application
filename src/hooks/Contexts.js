import { createContext } from 'react';

const IndexContext = createContext({
  activeIndex: null,
  setActiveIndex: () => {},
});

const KeyContext = createContext({
  activeKey: {
    eduActiveKey: null,
    expActiveKey: null,
  },
  setActiveKey: () => {},
});

const EduErrMsgContext = createContext({
  showEduErrMessage: false,
  setShowEduErrMessage: () => {},
});

const ExpErrMsgContext = createContext({
  showExpErrMessage: false,
  setShowExpErrMessage: () => {},
});

const PersonalInfoContext = createContext({
  personalInfo: {
    profilePicture: '',
    fullName: '',
    currentJobTitle: '',
    email: '',
    mobile: '',
    city: '',
  },
  setPersonalInfo: () => {},
});

const SummaryContext = createContext({
  summary: '',
  setSummary: () => {},
});

const ExperienceContext = createContext({
  experiences: {
    id: {
      jobTitle: '',
      employer: '',
      jobStartDate: '',
      jobEndDate: '',
      jobDescription: '',
    },
  },
  setExperiences: () => {},
});

const EducationContext = createContext({
  educations: {
    id: {
      degree: '',
      university: '',
      eduStartDate: '',
      eduEndDate: '',
      eduInfo: '',
    },
  },
  setEducations: () => {},
});

export {
  EducationContext,
  EduErrMsgContext,
  ExperienceContext,
  ExpErrMsgContext,
  IndexContext,
  KeyContext,
  PersonalInfoContext,
  SummaryContext,
};
