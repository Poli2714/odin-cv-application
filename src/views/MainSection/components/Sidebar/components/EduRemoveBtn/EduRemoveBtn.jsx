import { useContext } from 'react';

import { RemoveBtn } from 'src/components/buttons';
import { EducationContext, KeyContext } from 'src/hooks/Contexts';

function EduRemoveBtn() {
  const { educations, setEducations } = useContext(EducationContext);
  const { activeKey, setActiveKey } = useContext(KeyContext);

  const handleClickRemoveEdu = key => {
    const newEducations = { ...educations };
    delete newEducations[key];
    setEducations(newEducations);
    setActiveKey({
      ...activeKey,
      eduActiveKey: Object.keys(newEducations).at(-1),
    });
  };

  return (
    <RemoveBtn
      disabled={Object.keys(educations).length === 1 || !activeKey.eduActiveKey}
      onClick={() => handleClickRemoveEdu(activeKey.eduActiveKey)}
    />
  );
}

export default EduRemoveBtn;
