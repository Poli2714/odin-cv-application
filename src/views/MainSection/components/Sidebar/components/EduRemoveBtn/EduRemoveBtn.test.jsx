import { expect, test, vi } from 'vitest';
import { render, screen } from '@testing-library/react';

import EduRemoveBtn from './EduRemoveBtn';
import { EducationContext, KeyContext } from 'src/hooks/Contexts';

vi.mock('src/components/buttons', () => ({
  RemoveBtn: vi.fn(({ disabled }) => (
    <button disabled={disabled}>Mock click</button>
  )),
}));

function MockParentComponent({ activeKey, educations }) {
  return (
    <EducationContext.Provider value={{ educations }}>
      <KeyContext.Provider value={{ activeKey }}>
        <EduRemoveBtn />
      </KeyContext.Provider>
    </EducationContext.Provider>
  );
}

test('renders EduRemoveBtn', () => {
  render(<EduRemoveBtn />);

  expect(screen.getByRole('button')).toMatchSnapshot();
});

test('disables RemoveBtn if there is only one education', async () => {
  const educations = { id: { someData: '' } };
  const activeKey = { eduActiveKey: 'key' };

  render(<MockParentComponent activeKey={activeKey} educations={educations} />);
  const button = screen.getByRole('button');

  expect(button).toBeDisabled();
});

test('disables RemoveBtn if there are more than one experience and all are collapsed (not displayed)', () => {
  const educations = {
    id: { someData: 'foo' },
    id2: { someOtherData: 'bar' },
  };
  const activeKey = { eduActiveKey: '' };

  render(<MockParentComponent activeKey={activeKey} educations={educations} />);
  const button = screen.getByRole('button');

  expect(button).toBeDisabled();
});

test('enables RemoveBtn if there are more than one experience and one of them is expanded (displayed)', () => {
  const educations = {
    id: { someData: 'foo' },
    id2: { someOtherData: 'bar' },
  };
  const activeKey = { eduActiveKey: 'hey' };

  render(<MockParentComponent activeKey={activeKey} educations={educations} />);
  const button = screen.getByRole('button');

  expect(button).toBeEnabled();
});
