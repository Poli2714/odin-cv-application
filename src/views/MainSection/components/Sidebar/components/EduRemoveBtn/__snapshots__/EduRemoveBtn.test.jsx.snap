// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders EduRemoveBtn 1`] = `
<button
  disabled=""
>
  Mock click
</button>
`;
